### PR TITLE
Expose move_into_folder on the engine

### PR DESCRIPTION
**Pull request by d-m-u**
expose ```move_into_folder``` on the engine.

@miq-bot add_reviewer @lfu 
@miq-bot assign @tinaafitz 

For https://bugzilla.redhat.com/show_bug.cgi?id=1716858

# depends on
https://github.com/ManageIQ/manageiq/pull/19086

### CODE REVIEW

**Comment by d-m-u** [comment]
@miq-bot add_label enhancement, ivanchuk/yes

File: lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb [hunk 1]
```diff
@@ -15,5 +15,10 @@ def add_disk(disk_name, disk_size_mb, options = {})
     def remove_disk(disk_name, options = {})
       sync_or_async_ems_operation(options[:sync], "remove_disk", [disk_name, options])
     end
+
+    def move_into_folder(folder, options = {})
```
Comment (lfu) [inline lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb]: Is options needed here?

Comment (d-m-u) [reply lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb]: Well, it takes options -- https://github.com/ManageIQ/manageiq-providers-vmware/blob/78e3358127cc3542f8be1e3de9f7963759135d6a/app/models/manageiq/providers/vmware/infra_manager.rb#L265

File: lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb [hunk 2]
```diff
@@ -15,5 +15,10 @@ def add_disk(disk_name, disk_size_mb, options = {})
     def remove_disk(disk_name, options = {})
       sync_or_async_ems_operation(options[:sync], "remove_disk", [disk_name, options])
     end
+
+    def move_into_folder(folder)
+      raise ArgumentError, "folder must be a MiqAeServiceEmsFolder" unless folder.kind_of?(MiqAeMethodService::MiqAeServiceEmsFolder)
+      ar_method { @object.move_into_folder(EmsFolder.find_by(:id => folder.id)) }
```
Comment (gmcculloug) [inline lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb]: The call to VMware is here and is synchronous so we should wrap this in the same `sync_or_async_ems_operation` method as the above methods which will require the (initially added, then removed) `options` hash parameter.

See `moveIntoFolder` method: https://github.com/ManageIQ/vmware_web_service/blob/501b80924d89573aed8782aa59a4f6b75624a64c/lib/VMwareWebService/MiqVimFolder.rb#L84

File: lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb (hunk 1 above)
Comment (lfu) [reply lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb]: It is calling [this method](https://github.com/ManageIQ/manageiq/blob/23a1cc4ef6be627e3e4deeeb7a908923fd020150/app/models/vm_or_template/operations/relocation.rb#L109).

**Comment by d-m-u** [comment]
marked as wip because no queuing ar objects

**Comment by d-m-u** [comment]
@miq-bot remove_label wip

File: lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb [hunk 3]
```diff
@@ -15,5 +15,9 @@ def add_disk(disk_name, disk_size_mb, options = {})
     def remove_disk(disk_name, options = {})
       sync_or_async_ems_operation(options[:sync], "remove_disk", [disk_name, options])
     end
+
+    def move_into_folder(folder, options = {})
+      sync_or_async_ems_operation(options[:sync], "move_into_folder", [folder.ems_ref])
```
Comment (gmcculloug) [inline lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-vmware-infra_manager-vm.rb]: We should keep the check you had for EmsFolder type as many objects expose the `ems_ref` property and could lead to difficult to debug errors.  Would be nice to deal with an invalid object type before queuing the work.

**Comment by d-m-u** [comment]
Anyone want to look at this again please?

**Comment by simaishi** [comment]
Ivanchuk backport details:

```
$ git log -1
commit 960830d3f106cd728319a3a677802659d43dc716
Author: Greg McCullough <gmccullo@redhat.com>
Date:   Mon Aug 12 12:05:25 2019 -0400

    Merge pull request #344 from d-m-u/expose_move_into_folder
    
    Expose move_into_folder on the engine
    
    (cherry picked from commit 445f6a0505eb025638853be5a59f41832cb1fa85)
    
    https://bugzilla.redhat.com/show_bug.cgi?id=1716858
```

**Pull request merged**